### PR TITLE
feat(oxc): add `Compiler` and `CompilerInterface`

### PR DESCRIPTION
**Pull request by Boshen**
This PR adds a full compiler pipeline to the `oxc` crate, to stop us from implementing the same pipeline over and over again.

relates #4455

### CODE REVIEW

File: tasks/coverage/parser_babel.snap [hunk 1]
```diff
@@ -3,7 +3,7 @@ commit: 12619ffe
 parser_babel Summary:
 AST Parsed     : 2093/2101 (99.62%)
 Positive Passed: 2083/2101 (99.14%)
-Negative Passed: 1380/1501 (91.94%)
+Negative Passed: 1367/1501 (91.07%)
```
Comment (Boshen) [inline tasks/coverage/parser_babel.snap:6]: I'll investigate when I wake up.

File: crates/oxc/src/compiler.rs
Comment (DonIsaac) [inline crates/oxc/src/compiler.rs:1]: How are we going to handle multi-file outputs, such as source maps and ID?

File: crates/oxc/src/compiler.rs [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+        &self,
+        program: &Program<'a>,
+        source_text: &'a str,
+        source_type: SourceType,
+        trivias: &Trivias,
+    ) -> SemanticBuilderReturn<'a> {
+        SemanticBuilder::new(source_text, source_type)
+            .with_trivias(trivias.clone())
+            .with_check_syntax_error(true)
+            // .build_module_record(source_path.to_path_buf(), &program)
```
Comment (DonIsaac) [inline crates/oxc/src/compiler.rs:177]: This would be very useful to have

File: tasks/coverage/parser_test262.snap [hunk 3]
```diff
@@ -3,11 +3,18 @@ commit: a1587416
 parser_test262 Summary:
 AST Parsed     : 45859/45859 (100.00%)
 Positive Passed: 45859/45859 (100.00%)
-Negative Passed: 3925/3929 (99.90%)
+Negative Passed: 3918/3929 (99.72%)
```
Comment (DonIsaac) [inline tasks/coverage/parser_test262.snap:6]: What happened here?

File: crates/oxc/src/compiler.rs
Comment (Boshen) [reply crates/oxc/src/compiler.rs:1]: Not yet, this is for setting up the most basic things.

**Comment by DonIsaac** [comment]
This is a clean abstraction, I'm in favor of it.

**Pull request merged**